NFT/src/main.sw

```diff
@@ -0,0 +1,359 @@
+    /// This is only set on the initalization of the contract.
+    access_control: bool,
+
+    /// Stores an `Option` of the `Identity` that has permission to mint if `access_control` is set to true.
+    /// Will store `None` if this contract does not have `access_control` set.
+    /// Only the `admin` is allowed to change the `admin` of the contract.
+    admin: Option<Identity>,
+
+    /// Used of O(1) lookup of the number of tokens owned by each `Identity`.
+    /// This increments or decrements when minting, transfering ownership, and burning tokens.
+    /// Map(Identity => balance)
+    balances: StorageMap<Identity,
+    u64>, /// Stores the `Metadata` for each token based on the token's `u64` id
+    /// Map(token_id => Metadata)
+    meta_data: StorageMap<u64,
+    Option<MetaData>>, /// Maps a tuple of (owner, operator) identities and stores whether the
+    /// operator is allowed to transfer ALL tokens on the owner's behalf.
+    /// Map((owner, operator) => approved)
+    operator_approval: StorageMap<(Identity,
+    Identity), bool>, /// The total number of tokens that have been minted.
+    /// This should only be incremented.
+    token_count: u64,
+
+    /// The total supply tokens that can be minted. Can only be set on the initalization of the
+    /// contract.
+    token_supply: u64,
+}
+
+impl NFT for Contract {
+    /// Constructor for the NFT. Calling this function will instantiate the `total_supply`, the `admin`
+    /// `Identity`, and the `access_control` boolean. These values can only be set once.
+    /// Before this function is called, the contract is unable to perform any minting or transfering of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `admin` - The `Identity` which has the ability to mint if `access_control` is set to true and change the contract's admin.
+    /// * `access_control` - The `bool` which will determine whether identities will need to approval to mint.
+    /// * `token_supply` - The `u64` number representing the total supply of tokens which will be allowed to mint.
+    ///
+    /// # Reverts
+    ///
+    /// * When the constructor function has already been called.
+    /// * When the `token_supply` is set to 0.
+    /// * When `access_control` is set to true and no admin was given.
+    #[storage(read, write)]fn constructor(admin: Option<Identity>, access_control: bool, token_supply: u64) {
+        // This function can only be called once so if the token supply is already set it has
+        // already been called
+        require(storage.token_supply == 0, InitError::CannotReinitialize);
+        // The number of tokens that can be minted cannot be 0
+        require(token_supply != 0, InputError::TokenSupplyCannotBeZero);
+        // Access control is set to true but there was no admin given
+        require((access_control && admin.is_some()) || admin.is_none(), InitError::AccessControlSetAndAdminIsNone);
```

Comment:
Have you tested this with a case where `access_control` is true, but `admin.is_none()`?
It looks like the first '&&' condition will fail, but then the second would pass.

Reply:
Yes there is a test for this in the harness.rs